File: src/main/java/net/adventuregame/story/StoryManager.java
package net.adventuregame.story;

import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import com.mojang.serialization.DataResult;
import com.mojang.serialization.DynamicOps;
import net.adventuregame.game.GameState;

import java.util.*;

public class StoryManager {

    private StoryStage currentStage = StoryStage.THE_AWAKENING;
    private final Map<String, Quest> quests = new LinkedHashMap<>();
    private Dialogue currentDialogue = null;

    public StoryManager() {
        initQuests();
    }

    private void initQuests() {
        // Define quests per stage, you can expand these with real game logic conditions
        quests.put("find_lost_sword", new Quest(
                "find_lost_sword",
                "Find the legendary katana in the Forbidden Forest.",
                () -> {
                    // Replace this with your real check, e.g., player inventory or event flag
                    return GameState.player.hasItemByName("katana");
                }));

        quests.put("defeat_shin_shogun", new Quest(
                "defeat_shin_shogun",
                "Defeat the corrupted Shin Shogun and end the Shin Virus.",
                () -> {
                    // Replace with real boss defeated flag
                    return GameState.isBossDefeated("shin_shogun");
                }));
    }

    public StoryStage getCurrentStage() {
        return currentStage;
    }

    public void advanceStage() {
        switch (currentStage) {
            case THE_AWAKENING -> currentStage = StoryStage.THE_LOST_SWORD;
            case THE_LOST_SWORD -> currentStage = StoryStage.THE_SHIN_SHOGUN_RETURN;
            case THE_SHIN_SHOGUN_RETURN -> currentStage = StoryStage.THE_FINAL_CONFRONTATION;
            case THE_FINAL_CONFRONTATION -> currentStage = StoryStage.NEW_DAWN;
            case NEW_DAWN -> {
                // Story complete — maybe unlock new content or loop
            }
        }
        triggerStageEvents();
    }

    private void triggerStageEvents() {
        // Hook in your world event spawning, dialog triggers, cutscenes, etc.
        switch (currentStage) {
            case THE_AWAKENING -> {
                currentDialogue = new Dialogue("Village Elder", List.of(
                        "The prophecy awakens, young warrior...",
                        "Darkness stirs once more in our land."
                ));
            }
            case THE_LOST_SWORD -> {
                currentDialogue = new Dialogue("Old Samurai", List.of(
                        "You must find the lost katana deep within the Forbidden Forest.",
                        "Only then can you challenge the Shin Shogun."
                ));
            }
            case THE_SHIN_SHOGUN_RETURN -> {
                currentDialogue = new Dialogue("Messenger", List.of(
                        "The Shin Shogun grows stronger every day.",
                        "Prepare yourself for the coming battle."
                ));
            }
            case THE_FINAL_CONFRONTATION -> {
                currentDialogue = new Dialogue("Your Spirit", List.of(
                        "This is it. Face the darkness and end this curse!"
                ));
            }
            case NEW_DAWN -> {
                currentDialogue = new Dialogue("Village Elder", List.of(
                        "Peace has returned, but the world will always need heroes."
                ));
            }
        }
    }

    public Dialogue getCurrentDialogue() {
        return currentDialogue;
    }

    public void clearDialogue() {
        currentDialogue = null;
    }

    public Collection<Quest> getActiveQuests() {
        // Filter quests by current stage or all uncompleted quests
        // For simplicity, return all uncompleted quests
        List<Quest> active = new ArrayList<>();
        for (Quest q : quests.values()) {
            if (!q.isCompleted()) active.add(q);
        }
        return active;
    }

    // ======= Save / Load with Codec =======

    public static final Codec<StoryManager> CODEC = RecordCodecBuilder.create(instance ->
            instance.group(
                    Codec.STRING.fieldOf("currentStage").forGetter(story -> story.currentStage.name()),
                    Codec.unboundedMap(Codec.STRING, QuestCodec.CODEC).fieldOf("quests").forGetter(story -> story.quests),
                    DialogueCodec.CODEC.optionalFieldOf("currentDialogue").forGetter(story -> Optional.ofNullable(story.currentDialogue))
            ).apply(instance, (stageName, questsMap, dialogueOpt) -> {
                StoryManager sm = new StoryManager();
                sm.currentStage = StoryStage.valueOf(stageName);

                sm.quests.clear();
                sm.quests.putAll(questsMap);
                sm.currentDialogue = dialogueOpt.orElse(null);
                return sm;
            })
    );

    public Quest getQuest(String id) {
        return quests.get(id);
    }


    // Helper Codec wrapper classes for Quest and Dialogue

    public static class QuestCodec implements Codec<Quest> {
        public final String id;
        public final String description;
        public final boolean completed;

        public QuestCodec(Quest quest) {
            this.id = quest.getId();
            this.description = quest.getDescription();
            this.completed = quest.isCompleted();
        }

        public static final Codec<Quest> CODEC = RecordCodecBuilder.create(instance ->
                instance.group(
                        Codec.STRING.fieldOf("id").forGetter(Quest::getId),
                        Codec.STRING.fieldOf("description").forGetter(Quest::getDescription),
                        Codec.BOOL.fieldOf("completed").forGetter(Quest::isCompleted)
                ).apply(instance, (id, description, completed) -> {
                    Quest q = new Quest(id, description, () -> false); // dummy condition, to be restored
                    q.setCompleted(completed);
                    return q;
                })
        );


        @Override
        public <T> DataResult<Pair<Quest, T>> decode(DynamicOps<T> dynamicOps, T t) {
            return null;
        }

        @Override
        public <T> DataResult<T> encode(Quest quest, DynamicOps<T> dynamicOps, T t) {
            return null;
        }
    }


    public static class DialogueCodec implements Codec<Dialogue> {
        public final String speaker;
        public final List<String> lines;
        public final int currentLine;

        public DialogueCodec(Dialogue dialogue) {
            this.speaker = dialogue.getSpeaker();
            this.lines = dialogue.lines;
            this.currentLine = dialogue.currentLine;
        }

        // Factory for codec deserialization
        public static final Codec<Dialogue> CODEC = RecordCodecBuilder.create(instance ->
                instance.group(
                        Codec.STRING.fieldOf("speaker").forGetter(d -> d.getSpeaker()),
                        Codec.STRING.listOf().fieldOf("lines").forGetter(d -> d.lines),
                        Codec.INT.fieldOf("currentLine").forGetter(d -> d.currentLine)
                ).apply(instance, (speaker, lines, currentLine) -> {
                    Dialogue d = new Dialogue(speaker, lines);
                    d.currentLine = currentLine;
                    return d;
                })
        );

        @Override
        public <T> DataResult<Pair<Dialogue, T>> decode(DynamicOps<T> dynamicOps, T t) {
            return null;
        }

        @Override
        public <T> DataResult<T> encode(Dialogue dialogue, DynamicOps<T> dynamicOps, T t) {
            return null;
        }
    }

}

File: src/main/kotlin/net/adventuregame/story/StoryManager.kt
package net.adventuregame.story

import com.adv.core.font.GUIText
import com.mojang.datafixers.util.Pair
import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.DynamicOps
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.adventuregame.cutscene.Cutscene
import net.adventuregame.cutscene.CutsceneLine
import net.adventuregame.game.AdventureGame
import net.adventuregame.game.GameState
import net.adventuregame.guis.GuiTexture
import net.adventuregame.player.Player
import org.joml.Vector2f
import java.util.*
import java.util.function.Function
import java.util.function.Supplier

class StoryManager {
    var currentStage: StoryStage = StoryStage.THE_AWAKENING
        private set
    private val quests: MutableMap<String?, Quest> = LinkedHashMap<String?, Quest>()
    var currentDialogue: Dialogue? = null
        private set

    val player: Player
        get() = GameState.getInstance().player

    fun addQuest(id: String, quest: Quest) {
        quests[id] = quest
        println("Quest Added: $id")
        println("Total Quests: ${quests.keys}")
    }

    private fun initQuests() {
        // Define quests per stage, you can expand these with real game logic conditions
        quests["find_lost_sword"] = Quest(
            "find_lost_sword",
            "Find the legendary katana in the Forbidden Forest."
        ) { player.hasItemByName("katana") }.apply {
            onComplete = { GameState.getInstance().playCutscene(Cutscene(
                null,
                listOf(
                    CutsceneLine(
                        "やった！刀を見つけた！!"
                    )
                )
            )) }
        }
    }

    fun giveQuest(questId: String) {
        val quest = getQuest(questId)
        if (quest != null && !quest.isCompleted()) {
            // Add to active quests (already handled by StoryManager via quests map)
            println("[QUEST ASSIGNED] ${quest.description}")
            player.currentQuest = quest
            activeQuests.add(quest)
        } else {
            println("[QUEST ERROR] Quest $questId not found or already completed")
        }
    }

    val activeQuests: MutableCollection<Quest> = mutableSetOf()

    init {
        initQuests()
    }

    fun getQuest(id: String?): Quest? {
        return quests[id]
    }


    // Helper Codec wrapper classes for Quest and Dialogue
    class QuestCodec(quest: Quest) : Codec<Quest?> {
        val id: String?
        val description: String?
        val completed: Boolean

        init {
            this.id = quest.id
            this.description = quest.description
            this.completed = quest.isCompleted()
        }

        override fun <T> decode(dynamicOps: DynamicOps<T?>?, t: T?): DataResult<Pair<Quest?, T?>?>? {
            return null
        }

        override fun <T> encode(quest: Quest?, dynamicOps: DynamicOps<T?>?, t: T?): DataResult<T?>? {
            return null
        }

        companion object {
            val CODEC: Codec<Quest?> =
                RecordCodecBuilder.create<Quest?>(Function { instance: RecordCodecBuilder.Instance<Quest?>? ->
                    instance!!.group<String?, String?, Boolean?>(
                        Codec.STRING.fieldOf("id").forGetter<Quest?> { obj: Quest? -> obj!!.id },
                        Codec.STRING.fieldOf("description")
                            .forGetter<Quest?> { obj: Quest? -> obj!!.description },
                        Codec.BOOL.fieldOf("completed")
                            .forGetter<Quest?> { obj: Quest? -> obj!!.isCompleted() }
                    ).apply<Quest?>(instance) { id: String?, description: String?, completed: Boolean? ->
                        val q = Quest(id, description) { false } // dummy condition, to be restored
                        q.setCompleted(completed!!)
                        q
                    }
                }
                )
        }
    }

    class DialogueCodec(dialogue: Dialogue) : Codec<Dialogue> {
        val speaker: String?
        val lines: MutableList<String?>?
        val currentLine: Int

        init {
            this.speaker = dialogue.speaker
            this.lines = dialogue.lines
            this.currentLine = dialogue.currentLine
        }

        override fun <T> decode(dynamicOps: DynamicOps<T?>?, t: T?): DataResult<Pair<Dialogue?, T?>?>? {
            return null
        }

        override fun <T> encode(dialogue: Dialogue?, dynamicOps: DynamicOps<T?>?, t: T?): DataResult<T?>? {
            return null
        }

        companion object {
            // Factory for codec deserialization
            val CODEC =
                RecordCodecBuilder.create<Dialogue?>(Function { instance: RecordCodecBuilder.Instance<Dialogue?>? ->
                    instance!!.group<String?, MutableList<String?>?, Int?>(
                        Codec.STRING.fieldOf("speaker")
                            .forGetter<Dialogue?>(Function { d: Dialogue? -> d!!.speaker }),
                        Codec.STRING.listOf().fieldOf("lines")
                            .forGetter<Dialogue?> { d: Dialogue? -> d!!.lines },
                        Codec.INT.fieldOf("currentLine")
                            .forGetter<Dialogue?> { d: Dialogue? -> d!!.currentLine }
                    ).apply<Dialogue?>(
                        instance
                    ) { speaker: String?, lines: MutableList<String?>?, currentLine: Int? ->
                        val d = Dialogue(speaker, lines!!)
                        d.currentLine = currentLine!!
                        d
                    }
                }
                )
        }
    }

    companion object {
        // ======= Save / Load with Codec =======
        val CODEC: Codec<StoryManager?> =
            RecordCodecBuilder.create<StoryManager?>(Function { instance: RecordCodecBuilder.Instance<StoryManager?>? ->
                instance!!.group<String?, MutableMap<String, Quest>, Optional<Dialogue?>?>(
                    Codec.STRING.fieldOf("currentStage")
                        .forGetter<StoryManager?> { story: StoryManager? -> story!!.currentStage.name },
                    Codec.unboundedMap<String, Quest>(Codec.STRING, QuestCodec.Companion.CODEC).fieldOf("quests")
                        .forGetter<StoryManager?> { story: StoryManager? -> story!!.quests },
                    DialogueCodec.CODEC.optionalFieldOf("currentDialogue")
                        .forGetter { story -> Optional.ofNullable(story?.currentDialogue) }
                ).apply<StoryManager?>(
                    instance
                ) { stageName: String?, questsMap: MutableMap<String, Quest>, dialogueOpt: Optional<Dialogue?>? ->
                    val sm = StoryManager()
                    sm.currentStage = StoryStage.valueOf(stageName!!)

                    sm.quests.clear()
                    sm.quests.putAll(questsMap)
                    sm.currentDialogue = dialogueOpt!!.orElse(null)
                    sm
                }
            }
            )
    }
}

